store/tikv/range_task.go

```diff
@@ -0,0 +1,298 @@
+)
+
+// RangeTaskRunner splits a range into many ranges to process concurrently, and convenient to send requests to all
+// regions in the range. Because of merging and splitting, it's possible that multiple requests for disjoint ranges are
+// sent to the same region.
+type RangeTaskRunner struct {
+	name            string
+	store           Storage
+	concurrency     int
+	handler         RangeTaskHandler
+	maxBackoff      int
+	statLogInterval time.Duration
+
+	completedRegions int32
+}
+
+// RangeTaskHandler is the type of functions that processes a task of a region. It should be able to be invoked multiple
+// times.
+// If the range can't be processed immediately, returns `nextKey` to indicate the next task will start from `nextKey`.
+// If `nextKey` is set, it's expected to be inner range `r`.
+type RangeTaskHandler = func(ctx context.Context, bo *Backoffer, r kv.KeyRange, loc *KeyLocation) (nextKey []byte, err error)
+
+// NewRangeTaskRunner creates a RangeTaskRunner.
+//
+// `requestCreator` is the function used to create RPC request according to the given range.
+// `responseHandler` is the function to process responses of errors. If `responseHandler` returns error, the whole job
+// will be canceled.
+func NewRangeTaskRunner(
+	name string,
+	store Storage,
+	concurrency int,
+	handler RangeTaskHandler,
+) *RangeTaskRunner {
+	return &RangeTaskRunner{
+		name:            name,
+		store:           store,
+		concurrency:     concurrency,
+		handler:         handler,
+		maxBackoff:      rangeTaskDefaultMaxBackoff,
+		statLogInterval: rangeTaskDefaultStatLogInterval,
+	}
+}
+
+// SetMaxBackoff sets the max backoff for each single task.
+func (s *RangeTaskRunner) SetMaxBackoff(backoff int) {
+	s.maxBackoff = backoff
+}
+
+// SetStatLogInterval sets the time interval to log the stats.
+func (s *RangeTaskRunner) SetStatLogInterval(interval time.Duration) {
+	s.statLogInterval = interval
+}
+
+// RunOnRange runs the task on the given range.
+func (s *RangeTaskRunner) RunOnRange(ctx context.Context, startKey []byte, endKey []byte) error {
+	if len(endKey) != 0 && bytes.Compare(startKey, endKey) >= 0 {
+		logutil.Logger(ctx).Info("empty range task executed. ignored",
+			zap.String("name", s.name),
+			zap.Binary("startKey", startKey),
+			zap.Binary("endKey", endKey))
+		return nil
+	}
+
+	logutil.Logger(ctx).Info("range task started",
+		zap.String("name", s.name),
+		zap.Binary("startKey", startKey),
+		zap.Binary("endKey", endKey))
+
+	s.completedRegions = 0
+
+	// Periodically log the progress
+	statLogTicker := time.NewTicker(s.statLogInterval)
+
+	ctx, cancel := context.WithCancel(ctx)
+	taskCh := make(chan *kv.KeyRange, s.concurrency)
+	var wg sync.WaitGroup
+
+	// Create workers that concurrently process the whole range.
+	for i := 0; i < s.concurrency; i++ {
+		w := s.createWorker(taskCh, &wg)
+		wg.Add(1)
+		go w.run(ctx, cancel)
+	}
+
+	defer func() {
+		close(taskCh)
+		statLogTicker.Stop()
+		cancel()
```

Comment:
last task may not handled. Worker may receive task from channel then quit due to cancel.